ocaml/xapi/extauth_plugin_ADwinbind.ml

```diff
@@ -547,6 +560,80 @@ let domainify_uname ~domain uname =
   else
     Printf.sprintf "%s@%s" uname domain
 
+let range s e step =
+  let rec aux n acc = if n >= e then acc else aux (n + step) (n :: acc) in
+  aux 0 [] |> List.rev
+
+let min_valid_pbis_value_length = String.length "X''"
+
+let extract_raw_value_from_pbis_db key =
+  let sql =
+    Printf.sprintf "select QUOTE(Value) from regvalues1 where ValueName='%s'"
+      key
+  in
+  let value =
+    Helpers.call_script ~log_output:On_failure !Xapi_globs.sqlite3
+      [pbis_db_path; sql]
+  in
+  if String.length value <= min_valid_pbis_value_length then
+    raise (Auth_service_error (E_GENERIC, Printf.sprintf "No value for %s" key))
+  else
+    value
+
+let remove_tailing s len = String.sub s 0 (String.length s - len)
```

Comment:
This may encounter problems if len is bigger than the length of s. I propose using 
```ocaml
Astring.String.drop ~rev:true ~max:len s
```